src/session_protocol.cpp

```diff
@@ -0,0 +1,651 @@
+        result.envelope.flags |= ENVELOPE_FLAGS_SOURCE_DEVICE;
+    }
+
+    // Parse server timestamp (optional)
+    if (envelope.has_servertimestamp()) {
+        result.envelope.server_timestamp = envelope.servertimestamp();
+        result.envelope.flags |= ENVELOPE_FLAGS_SERVER_TIMESTAMP;
+    }
+
+    // Parse content
+    if (!envelope.has_content())
+        throw std::runtime_error{"Parse decrypted message failed, missing content"};
+
+    // Decrypt content
+    // The envelope is encrypted in GroupsV2, contents unencrypted. In 1o1 and legacy groups, the
+    // envelope is encrypted, contents is encrypted.
+    bool has_encrypted_content = !keys.use_group_keys;
+    if (has_encrypted_content) {
+        std::span<const uint8_t> content_str = session::to_span(envelope.content());
+        auto [content_plaintext, sender_ed25519_pubkey] =
+                session::decrypt_incoming(keys.recipient_ed25519_privkey, content_str);
+        assert(result.sender_ed25519_pubkey.size() == crypto_sign_ed25519_PUBLICKEYBYTES);
+
+        result.content_plaintext = std::move(content_plaintext);
+        std::memcpy(
+                result.sender_ed25519_pubkey.data(),
+                sender_ed25519_pubkey.data(),
+                result.sender_ed25519_pubkey.size());
+
+        if (crypto_sign_ed25519_pk_to_curve25519(
+                    result.sender_x25519_pubkey.data(), result.sender_ed25519_pubkey.data()) != 0)
+            throw std::runtime_error(
+                    "Parse content failed, ed25519 public key could not be converted to x25519 "
+                    "key.");
+    } else {
+        result.content_plaintext.resize(envelope.content().size());
+        std::memcpy(
+                result.content_plaintext.data(),
+                envelope.content().data(),
+                envelope.content().size());
+    }
+
+    // TODO: We parse the content in libsession to extract pro metadata but we return the unparsed
+    // blob back to the caller. This is temporary, eventually we will return a proxy structure for
+    // the protobuf Content type to the user. We avoid returning the direct protobuf type to keep
+    // the interface simple and avoid leaking protobuf implementation detail into the libsession
+    // interface.
+    SessionProtos::Content content = {};
+    if (!content.ParseFromArray(result.content_plaintext.data(), result.content_plaintext.size()))
+        throw std::runtime_error{fmt::format(
+                "Parse content from envelope failed: {}", result.content_plaintext.size())};
+
+    // A signature must always be present on the envelope. This is to make a pro and non-pro
+    // envelope indistinguishable. If the message does not have pro then this signature must still
+    // be set but will be ignored. So in all instances a signature must be attached (real or
+    // dummy).
+    if (!envelope.has_prosig())
+        throw std::runtime_error("Parse envelope failed, message is missing pro signature");
```

Comment:
We can't do this, yet. This would throw for any messages already stored in the swarm, and until everyone has upgraded their app, right?